### PR TITLE
Code extraction performance improvements

### PR DESCRIPTION
**Pull request by imjoehaines**
## Goal

This PR improves the performance of extracting code from files in a stacktrace

The performance improvements are especially relevant when a stacktrace has the same file multiple times (e.g. when chaining private methods) and when the files are long

Exactly how much better this is is impossible to say without looking at a specific stacktrace — it can be more than a 3x speedup if it mostly consists of the same few files and they are very long, but every stacktrace is different 🌈. I didn't observe any noticable slow down when a stackrace contained exclusively different files, but the new code does involve a bit more work in the worst case scenario

### Added

- Code extraction is now handled by a `CodeExtractor` class. This exists to keep track of which files in the stacktrace we need to get code from (we will sometimes remove files from the processed stacktrace) and so that it can do a single pass per file, extracting all the regions of code that it needs in one go
- The `Stacktrace`class is now a module as it was always basically one method and was immediately converted to an array when it was called
- A bunch of new tests and fixtures for code extraction and some untested parts of the stacktrace processing

### Linked issues

Related to #481

## Review

For the submitter, initial self-review:

- [x] Commented on code changes inline explain the reasoning behind the approach
- [x] Reviewed the test cases added for completeness and possible points for discussion
- [ ] A changelog entry was added for the goal of this pull request
- [x] Check the scope of the changeset - is everything in the diff required for the pull request?
- This pull request is ready for:
  - [ ] Initial review of the intended approach, not yet feature complete
  - [ ] Structural review of the classes, functions, and properties modified
  - [x] Final review

For the pull request reviewer(s), this changeset has been reviewed for:

- [ ] Consistency across platforms for structures or concepts added or modified
- [ ] Consistency between the changeset and the goal stated above
- [ ] Internal consistency with the rest of the library - is there any overlap between existing interfaces and any which have been added?
- [ ] Usage friction - is the proposed change in usage cumbersome or complicated?
- [ ] Performance and complexity - are there any cases of unexpected O(n^3) when iterating, recursing, flat mapping, etc?
- [ ] Concurrency concerns - if components are accessed asynchronously, what issues will arise
- [ ] Thoroughness of added tests and any missing edge cases
- [ ] Idiomatic use of the language

### CODE REVIEW

File: lib/bugsnag/stacktrace.rb [hunk 1]
```diff
@@ -10,105 +11,63 @@ class Stacktrace
...
+          trace_hash.delete(:inProject) if file.match(configuration.vendor_path)
         end
 
-
         # Strip common gem path prefixes
         if defined?(Gem)
           file = Gem.path.inject(file) {|line, path| line.sub(/#{path}\//, "") }
         end
 
+        next if !file || file.empty?
```
Comment (kattrali) [inline lib/bugsnag/stacktrace.rb:54]: What case is this handling? The file path was equal to the project root / a gem path?

Comment (imjoehaines) [reply lib/bugsnag/stacktrace.rb:54]: Yeah exactly — [this test triggers it](https://github.com/bugsnag/bugsnag-ruby/blob/0fe026245b5e94939a3a407b856d8e1790dff56d/spec/stacktrace_spec.rb#L404-L425)

Comment (kattrali) [reply lib/bugsnag/stacktrace.rb:54]: Right but is that a thing that can happen? Backtrace lines that point to directories? If so, in the case that a line has a (non project root / gem path) directory, that should be something we think about in a few places.

Comment (imjoehaines) [reply lib/bugsnag/stacktrace.rb:54]: I don't think so, no. You could trigger it by setting the project root to a file rather than a directory too, which is a _little_ more realistic, but I don't think this is ever really triggered in practice

The only reason I kept it is because it [was here before](https://github.com/bugsnag/bugsnag-ruby/blob/59254960b973cbb8f635c29fe18526663b6c6a2c/lib/bugsnag/stacktrace.rb#L60-L64) and I can't say for certain that there's no reason for it. I went back through the history again and it was [implemented in this commit](https://github.com/bugsnag/bugsnag-ruby/commit/a479b6ed10ffdd3b808883037db92b27553aec6a) which also let users add their own filters for the stacktrace file name, so it would have been necessary then but seems like it's not anymore (unless the project root/a gem path point to a file)

Comment (kattrali) [reply lib/bugsnag/stacktrace.rb:54]: Hmmm ghosts in the code. I'd also suspect that its not needed anymore, especially since there are simpler mechanisms now for filtering out stacktrace lines if needed.

Comment (imjoehaines) [reply lib/bugsnag/stacktrace.rb:54]: spooky

👍 I'll remove it
